### PR TITLE
Draft: Fetch Hunks on demand

### PR DESCRIPTION
**Pull request by Murmele**
do not fetch all hunks at once, because for large files with a lot of hunks (example: Kicad Board file) will hang up Gittyup

### CODE REVIEW

**Comment by Murmele** [comment]
@stefanknotzer can you review this one?

**Comment by stefanknotzer** [comment]
Yes, I'll do a review

**Comment by stefanknotzer** [comment]
I am sorry: can't do a review.

The patch is not working for me. My test was to select the very first commit in the Gittyup repository and select the "dep" folder. Then I tried to get to the bottom of the HunkWidget by mouse wheel and scrollbar --> no chance.
Suggestions:
- Do not change or add anything to the FileWidget (makes things more complicated)
- A behavior like the CommitList scrollbar would be nice (load more hunks when the scrollbar hits the bottom)
- A QTimer might be helpful to load hunk by hunk till the scrollbar is visible / not at the bottom any more - iirc the hunks are loaded in the paint routine which causes delays in scrollbar events

**Comment by Murmele** [comment]
> 

I had a bug in it. For some reason mHunkLayout.children().count() did not represent the count of the hunks. It works better now, but I think it is more complicated than expected, because when discarding or staging single lines it might get complicated. I have to check it out.

- What do you mean with a behaviour like the Commit list?
I tried to just adding new hunks when they are needed in the diffview, as it is done for the FileWidget.

But how should I handle FileWidgets with more than 100Hunks (I had a Filewidget with more than 500 hunks in it, after I changed something in a KiCad file)?

**Comment by Murmele** [comment]
Maybe loading all hunks, but not creating the editor widget and creating the editor widget only when needed?

**Comment by stefanknotzer** [comment]
The patch is working!
I now understand that there are 2 performance issuses/problems:
1. Selecting a folder with many modified files (like the "dep" folder in the initial commit)
2. Selecting a file with countless hunks (I admit I never had or thought of that case before)

Give me a few days, I'll have a closer look, but I think your approach is fine (DiffView: load files/editors on demand, FileWidget: load hunks on demand)

File: src/ui/DiffView/FileWidget.cpp [hunk 1]
```diff
@@ -453,6 +483,7 @@ HunkWidget *FileWidget::addHunk(
 
 void FileWidget::stageHunks(const HunkWidget* hunk, git::Index::StagedState stageState, bool completeFile, bool completeFileStaged)
 {
+	// Might be a problem if not all hunks are loaded
```
Comment (Murmele) [inline src/ui/DiffView/FileWidget.cpp:486]: This will make problems

**Comment by Murmele** [comment]
No it is not working perfectly. I pushed also some comments, we have to check everywhere it will be looped over the hunks. Staging single lines and discarding will not work properly.
Idea: maybe writing the staged patch

**Comment by Murmele** [comment]
At the moment when staging a single line I iterate over all hunks and write the data back. When not loading all hunks we have to overwrite only the first n lines and let the others as they are. 
More complicated then expected, but I think it will be a good performance boost

File: src/ui/DiffView/DiffView.cpp [hunk 2]
```diff
@@ -381,6 +381,15 @@ void DiffView::fetchMore()
   QList<QModelIndex> indices = mDiffTreeModel->modelIndices(dtw->selectedIndex());
   int count = indices.count();
 
+  // First load all hunks of last file and then go on with loading new files
+  if (mFiles.count() > 0) {
+      auto lastFile = mFiles[mFiles.count() -1];
```
Comment (stefanknotzer) [inline src/ui/DiffView/DiffView.cpp:386]:
```suggestion
  if (!mFiles.isEmpty()) {
      auto lastFile = mFiles.last();
```

**Comment by stefanknotzer** [comment]
My personal opinion to get DiffView back on track:
1. Strip it to the core. Remove everything and only show 1 selected file (like the BlameEditor does). Do not care about multiple files for now. Remove the disclosure button and the context menu.
2. Make it pretty. Show icons for binary files (QFileIconProvider) and some additional info like file size and access rights. Maybe display pictures with fixed size (e.g. 128 x 128 pixel).
3. Implement merging on a hunk basis for text files and on file basis for binary files

I already coded 2. and 3. for GitAhead but never made a pull request.

**Comment by Murmele** [comment]
But how would you handle multiple files? I tested a few days ago, doing the calculation and setting the flags but not creating the editor and it seems to be fine. So maybe we can load the complete hunk, but not doing the assignement to the editor only when neccesary. In this case even staging of single line is working again. I have to test it next days.

Just create a merge request for the 2 and 3. Not sure if I understood the 3 correctly

**Comment by stefanknotzer** [comment]
For the start I would not handle multiple files in the DiffView. The DiffView should only show the one file that is selected in one of the TreeViews. If a folder (with more than one file) is selected in the TreeView, then the DiffView simply shows nothing - that would be the same behavior as the BlameEditor. This is definitely my preferred behavior over senseless scrolling down the hunks and getting lost because you can't remember which file is actually shown.

For merging I created a ResolutionWidget (based on the HunkWidget):
Conflict
Select "ours" or "theirs" and resolve the conflict
Merging
Save the resolution or undo the selection

**Comment by stefanknotzer** [comment]
I agree: an option for a simple DiffView (maybe a global setting) is a good solution. Might also help people with performance issues.
I always used an external merge tool or even another git client (like tortoise git) because the internal merge is awful. After fixing some core issues my last TODO would be to develop a good, easy to use internal merge solution.

**Comment by Murmele** [comment]
I agree with you, the current merge tool is not that powerful, but at least you can do right click on the file and ooen with a merge tool. So I am using mostly kdiff3 ans it works fine. So for me the priority of a new merge tool is quite low.

**Comment by Murmele** [comment]
but I did not understand whats the different of your resolution widget and the hunkwidget? the hunkwidget handles also theirs and ours.

**Comment by stefanknotzer** [comment]
My ResolutionWidget would be a real editor with all the plugins enabled (whitespace check, line length check, ...). So you can actually edit the merged solution and fix minor issues before committing.
My priority on that feature is also very low because the external merge tools are very useful (my favorite is Beyond Compare).

File: src/ui/DiffView/FileWidget.cpp (hunk 1 above)
Comment (stefanknotzer) [reply src/ui/DiffView/FileWidget.cpp:486]:
```suggestion
  fetchAll(-1);
```

**Comment by stefanknotzer** [comment]
If you want to see how the final version works, have a look at:
https://github.com/stefanknotzer/Gittyup/tree/release_20220205

Please feel free to cherry-pick what you like as I am breaking the changes down to simple pull-requests.

**Comment by Murmele** [comment]
> If you want to see how the final version works, have a look at: https://github.com/stefanknotzer/Gittyup/tree/release_20220205
> 
> Please feel free to cherry-pick what you like as I am breaking the changes down to simple pull-requests.

@stefanknotzer can I close than this PR and you open a new one?

**Pull request closed without merging**